docs/language/learn-ql/ql-etudes/river-crossing.rst

```diff
@@ -0,0 +1,244 @@
+River crossing puzzle
+#####################
+
+The aim of this tutorial is to write a QL query that finds a solution to the following classical logic puzzle:
+
+.. pull-quote::
+
+   River crossing puzzle
+
+   A man is trying to ferry a goat, a cabbage, and a wolf across a river.
+   His boat can only take himself and at most one item as cargo.
+   His problem is that if the goat is left alone with the cabbage, it will eat it.
+   And if the wolf is left alone with the goat, it will eat it.
+   How does he get everything across the river?
+
+A solution should be a set of instructions for how to ferry the items, such as "First ferry the goat
+across the river, and come back with nothing. Then ferry the cabbage across, and come back with ..."
+
+There are lots of ways to approach this problem and implement it in QL. Before you start, make
+sure that you are familiar with how to define `classes <https://help.semmle.com/QL/ql-handbook/classes.html>`__
+and `predicates <https://help.semmle.com/QL/ql-handbook/predicates.html>`__ in QL.
+The following walkthrough is just one of many possible implementations, so have a go at writing your
+own query too! To find more example queries, see the list :ref:`below <alternatives>`.
+
+Walkthrough
+-----------
+
+Model the elements of the puzzle
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The basic components of the puzzle are the cargo items and the shores on either side of the river.
+Start by modeling these as classes.
+
+First, define a class ``Cargo`` containing the different cargo items.
+Note that the man can also travel on his own, so it helps to explicitly include ``"Nothing"`` as
+a piece of cargo.
+
+.. container:: toggle
+
+   .. container:: name
+
+      *Show/hide code*
+
+   .. literalinclude:: river-crossing.ql
+      :lines: 15-22
+
+Second, any item can be on one of two shores. Let's call these the "left shore" and the "right shore".
+Define a class ``Shore`` containing ``"Left"`` and ``"Right"``.
+
+It would be helpful to express "the other shore". You can do this by defining a member predicate
```

Comment:
```suggestion
It would be helpful to express "the other shore" to model moving from one side of the river to the other. You can do this by defining a member predicate
```